lib/compiler_rt/x86_cpu_model.zig

```diff
@@ -0,0 +1,428 @@
+    };
+
+    const features = blk: {
+        var detected = Target.Cpu.Feature.Set.empty;
+        x86.detectNativeFeatures(&detected, builtin.os.tag);
+
+        var features = std.mem.zeroes(FeatureSet);
+        inline for (@typeInfo(Target.x86.Feature).Enum.fields) |f| {
+            const index: Target.Cpu.Feature.Set.Index = f.value;
+            if (comptime !@hasField(Feature, f.name)) continue;
+            if (detected.isEnabled(index)) setFeature(&features, @field(Feature, f.name));
+        }
+
+        if (hasFeature(features, .@"64bit") and hasFeature(features, .sse2)) {
+            setFeature(&features, .x86_64_baseline);
+
+            if (hasFeature(features, .cx16) and hasFeature(features, .popcnt) and
+                hasFeature(features, .sahf) and hasFeature(features, .sse4_2))
+            {
+                setFeature(&features, .x86_64_v2);
+
+                if (hasFeature(features, .avx2) and hasFeature(features, .bmi) and
+                    hasFeature(features, .bmi2) and hasFeature(features, .f16c) and
+                    hasFeature(features, .fma) and hasFeature(features, .lzcnt) and
+                    hasFeature(features, .movbe))
+                {
+                    setFeature(&features, .x86_64_v3);
+
+                    if (hasFeature(features, .avx512bw) and hasFeature(features, .avx512cd) and
+                        hasFeature(features, .avx512dq) and hasFeature(features, .avx512vl))
+                    {
+                        setFeature(&features, .x86_64_v4);
+                    }
+                }
+            }
+        }
+
+        break :blk features;
+    };
+
+    cpu.vendor = switch (vendor) {
+        .intel => blk: {
+            cpu.type, cpu.subtype = getIntelTypeAndSubtype(family, model, features);
+            break :blk .intel;
+        },
+        .amd => blk: {
+            cpu.type, cpu.subtype = getAmdTypeAndSubtype(family, model, features);
+            break :blk .amd;
+        },
+        else => .other,
+    };
+    cpu.features[0] = features[0];
+    cpu_extra_features = features[1..].*;
+
+    return 0;
+}
+
+const Model = extern struct {
+    vendor: Vendor = .unknown,
+    type: Type = .unknown,
+    subtype: Subtype = .unknown,
+    features: [1]u32 = .{0},
+};
+
+const FeatureSet = [4]u32;
+comptime {
+    inline for (@typeInfo(Feature).Enum.fields) |f| {
+        if (f.value >= @bitSizeOf(FeatureSet))
+            @compileError(@import("std").fmt.comptimePrint("Feature.{s} ({}) bitindex too large", .{ f.name, f.value }));
+    }
+}
+
+fn hasFeature(set: FeatureSet, f: Feature) bool {
+    const f_value = @intFromEnum(f);
+    const bit = set[f_value / 32] & (@as(u32, 1) << @as(u5, @intCast(f_value % 32)));
+    return bit != 0;
+}
+fn setFeature(set: *FeatureSet, f: Feature) void {
+    const f_value = @intFromEnum(f);
+    set[f_value / 32] |= (@as(u32, 1) << @as(u5, @intCast(f_value % 32)));
+}
+
+// These values match the values in LLVM.
```

Comment:
Can we make checks for this to detect divergence from LLVM? Maybe via filtering the target feature set?

Reply:
I was wondering, if something like https://github.com/ziglang/zig/blob/8a8fd47d217eadb8a45ec6b9325acfe1d42704d7/src/zig_llvm.cpp#L1174, ie as being part of https://github.com/ziglang/zig/blob/8a8fd47d217eadb8a45ec6b9325acfe1d42704d7/tools/update_cpu_features.zig would be feasible to keep things synchronized with upstream.

I don't think this is a requirement to merge, just something nice to have for maintenance.

Reply:
Ohhh I see that's how Zig generates the `std.Target.x86` enums. Unfortunately the compiler-rt feature enums seem to be completely separate from the features defined in `llvm/lib/Target/X86/X86.td` which is why the upstream implementation of cpu_model looks kind of sloppy with different feature names or unused features.